### PR TITLE
[SYCLomatic] Re-implement query-api-mapping feature.

### PR DESCRIPTION
**Pull request by tangjj11**
Signed-off-by: Tang, Jiajun jiajun.tang@intel.com

### CODE REVIEW

**Comment by tangjj11** [comment]
The old PR #1115 is closed by mistake and cannot be reopened.

File: clang/lib/DPCT/SaveNewFiles.cpp [hunk 1]
```diff
@@ -457,26 +460,17 @@ int saveNewFiles(clang::tooling::RefactoringTool &Tool, StringRef InRoot,
...
-        std::string ErrMsg =
-            "[ERROR] Create file: " + std::string(OutPath.str()) + " fail.\n";
-        PrintMsg(ErrMsg);
-        status = MigrationSaveOutFail;
-        return status;
+      if (!DpctGlobalInfo::isQueryAPIMapping()) {
+        std::error_code EC;
+        EC = fs::create_directories(path::parent_path(OutPath));
+        if ((bool)EC) {
+          std::string ErrMsg =
```
Comment (yihwang-nv) [inline clang/lib/DPCT/SaveNewFiles.cpp]: We can use `llvm::Twine` to boost the performance of string concatenation.
Eg.
```
std::string ErrMsg = llvm::Twine("[ERROR] Create file :").concat(OutPath).concat(" fail: ").concat(EC.message()).concat('\n').str()
```

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp [hunk 2]
```diff
@@ -6,37 +6,32 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "llvm/Support/raw_ostream.h"
+
```
Comment (yihwang-nv) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: Extra blank line

**Review by yihwang-nv — APPROVED** [review]
LGTM % comments

File: clang/lib/DPCT/SaveNewFiles.cpp [hunk 3]
```diff
@@ -30,7 +31,9 @@
 
 #include "Utility.h"
 #include "llvm/Support/raw_os_ostream.h"
+#include "llvm/Support/raw_ostream.h"
```
Comment (yihwang-nv) [inline clang/lib/DPCT/SaveNewFiles.cpp]: Can we merge these LLVM header before L22? and seems this include order is not looks well.

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/SaveNewFiles.cpp (hunk 1 above)
Comment (tangjj11) [reply clang/lib/DPCT/SaveNewFiles.cpp]: Fallback all the changes in this file.

File: clang/lib/DPCT/SaveNewFiles.cpp (hunk 3 above)
Comment (tangjj11) [reply clang/lib/DPCT/SaveNewFiles.cpp]: Fallback all the changes in this file.

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp (hunk 2 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: done

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp [hunk 4]
```diff
@@ -7,36 +7,30 @@
...
 
-void APIMapping::registerEntry(const std::string &Name,
-                               const std::string &Description) {
-  if (Description == "NA")
-    return;
-  EntryMap[Name] = Description;
+void APIMapping::registerEntry(llvm::StringRef Name,
+                               llvm::StringRef SourceCode) {
+  if (EntryMap.contains(Name)) {
+    llvm::errs() << "The API Mapping cases name is duplicated.\n";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: Line 20 is not required for release build.   how about remove it in release build.

File: clang/lib/DPCT/CMakeLists.txt [hunk 5]
```diff
@@ -77,6 +77,20 @@ add_custom_target(dpct_helper_headers_and_inc
...
 
+# Handle API mapping cases header
+set(API_MAPPING_REGISTER_FOLDER ${CMAKE_BINARY_DIR}/tools/clang/include/clang/DPCT/)
+file(GLOB SRC_FILES ../../examples/DPCT/*/*.cu)
+file(WRITE ${API_MAPPING_REGISTER_FOLDER}APIMappingRegister.def "")
+foreach(FILE ${SRC_FILES})
+  file(READ ${FILE} FILE_STR)
+  string(REGEX REPLACE ".*/(.*).cu$" "\\1" API_NAME ${FILE})
+  file(APPEND ${API_MAPPING_REGISTER_FOLDER}APIMappingRegister.def
+  "registerEntry(\"" "${API_NAME}" "\",\nR\"(" "${FILE_STR}" ")\");\n")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/CMakeLists.txt:94]: how about add logic to check if ${API_NAME} is unique, if not, report error during build time.

File: clang/lib/DPCT/DPCT.cpp [hunk 6]
```diff
@@ -736,8 +739,33 @@ int runDPCT(int argc, const char **argv) {
...
+      llvm::outs() << "The API Mapping is not available\n";
+      dpctExit(MigrationSucceeded);
+    }
+    llvm::outs() << "CUDA API:" << getApiMappingStr(SourceCode);
+    int FD;
+    SmallString<64> TempFile;
+    if (auto EC =
+            llvm::sys::fs::createTemporaryFile("tmp", "cu", FD, TempFile)) {
+      llvm::errs() << "Error: dpct internal error. Cannot create temp file for "
+                      "api mapping.\n";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DPCT.cpp]:
```suggestion
                      "API Mapping.\n";
```

File: clang/lib/DPCT/DPCT.cpp [hunk 7]
```diff
@@ -736,8 +739,33 @@ int runDPCT(int argc, const char **argv) {
...
+    }
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DPCT.cpp:747]: if failed to create temp file, then exit the dpct program.

File: clang/lib/DPCT/DPCT.cpp [hunk 8]
```diff
@@ -736,8 +739,33 @@ int runDPCT(int argc, const char **argv) {
...
-                       OptParser->getSourcePathList());
+  std::vector<std::string> SourcePathList;
+  if (QueryAPIMapping.getNumOccurrences()) {
+    APIMapping::initEntryMap();
+    auto SourceCode = APIMapping::getAPISourceCode(QueryAPIMapping);
+    if (SourceCode.empty()) {
+      llvm::outs() << "The API Mapping is not available\n";
+      dpctExit(MigrationSucceeded);
+    }
+    llvm::outs() << "CUDA API:" << getApiMappingStr(SourceCode);
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DPCT.cpp]: let's delay the print, like after success create temp file.

File: clang/lib/DPCT/DPCT.cpp [hunk 9]
```diff
@@ -736,8 +739,33 @@ int runDPCT(int argc, const char **argv) {
...
 
-  RefactoringTool Tool(OptParser->getCompilations(),
-                       OptParser->getSourcePathList());
+  std::vector<std::string> SourcePathList;
+  if (QueryAPIMapping.getNumOccurrences()) {
+    APIMapping::initEntryMap();
+    auto SourceCode = APIMapping::getAPISourceCode(QueryAPIMapping);
+    if (SourceCode.empty()) {
+      llvm::outs() << "The API Mapping is not available\n";
+      dpctExit(MigrationSucceeded);
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DPCT.cpp]: let's define 2 error code for API mapping query:  1. API mapping not exist.  2. internal error (tmp file create fail)

File: clang/test/dpct/query_api_mapping/Math/test-after11.6.cu [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+// HFMA-NEXT:   __hfma(b1 /*__nv_bfloat16*/, b2 /*__nv_bfloat16*/, b3 /*__nv_bfloat16*/);
+// HFMA-NEXT: Is migrated to (with some neccessary option):
+// HFMA-NEXT:   sycl::ext::intel::math::hfma(h1, h2);
+// HFMA-NEXT:   b1 * b2 + b3;
+
+// RUN: dpct --cuda-include-path="%cuda-path/include" --query-api-mapping=__hfma_sat | FileCheck %s -check-prefix=HFMA_SAT
+// HFMA_SAT: CUDA API:
+// HFMA_SAT-NEXT:   __hfma_sat(h1 /*__half*/, h2 /*__half*/, h3 /*__half*/);
+// HFMA_SAT-NEXT:   __hfma_sat(b1 /*__nv_bfloat16*/, b2 /*__nv_bfloat16*/, b3 /*__nv_bfloat16*/);
+// HFMA_SAT-NEXT: Is migrated to (with some neccessary option):
```
Comment (zhimingwang36) [inline clang/test/dpct/query_api_mapping/Math/test-after11.6.cu]: user will be mislead by  " some necessary option",  how about list the options explicitly, (the option can be embed in test cases in example/ folder)

File: clang/test/dpct/query_api_mapping/Math/test-after11.6.cu [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+// RUN: dpct --cuda-include-path="%cuda-path/include" --query-api-mapping=__hfma | FileCheck %s -check-prefix=HFMA
+// HFMA: CUDA API:
+// HFMA-NEXT:   __hfma(h1 /*__half*/, h2 /*__half*/, h3 /*__half*/);
+// HFMA-NEXT:   __hfma(b1 /*__nv_bfloat16*/, b2 /*__nv_bfloat16*/, b3 /*__nv_bfloat16*/);
+// HFMA-NEXT: Is migrated to (with some neccessary option):
+// HFMA-NEXT:   sycl::ext::intel::math::hfma(h1, h2);
+// HFMA-NEXT:   b1 * b2 + b3;
+
+// RUN: dpct --cuda-include-path="%cuda-path/include" --query-api-mapping=__hfma_sat | FileCheck %s -check-prefix=HFMA_SAT
```
Comment (zhimingwang36) [inline clang/test/dpct/query_api_mapping/Math/test-after11.6.cu:12]: please opt the run case:  " dpct  --query-api-mapping" , it fails for no --cuda-include-path is specified,  please print msg to ask user to provide cuda-path.

File: clang/examples/DPCT/README.md [hunk 12]
```diff
@@ -0,0 +1,9 @@
+# How to add API mapping cases
+
+- Add the API mapping cases of different librerays in corresponding folder.
+
+- Make sure the file name is the same as the query API name.
```
Comment (zhimingwang36) [inline clang/examples/DPCT/README.md]:
```suggestion
- Make sure the file name is format as API_Name.cu.

```

File: clang/examples/DPCT/README.md [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+
```
Comment (zhimingwang36) [inline clang/examples/DPCT/README.md:6]: content of API_name.cu, please give an example, and for the example, please highlight the note on each part:
1. function define 
2. Arg list and comments. 
3. put calls between "// start" and "// end", all other code, don't put in this scope.
4. how and when "// option" is required.

File: clang/examples/DPCT/README.md [hunk 14]
```diff
@@ -0,0 +1,9 @@
...
+- Only the code between the `// Start` and the `// End` will be print when using `--query-api-mapping`.
+
+- Need format the code for better output.
```
Comment (zhimingwang36) [inline clang/examples/DPCT/README.md]: please format the code.

File: clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp (hunk 4 above)
Comment (tangjj11) [reply clang/lib/DPCT/APIMapping/QueryAPIMapping.cpp]: done

File: clang/lib/DPCT/CMakeLists.txt (hunk 5 above)
Comment (tangjj11) [reply clang/lib/DPCT/CMakeLists.txt:94]: done

File: clang/lib/DPCT/DPCT.cpp (hunk 8 above)
Comment (tangjj11) [reply clang/lib/DPCT/DPCT.cpp]: done

File: clang/lib/DPCT/DPCT.cpp (hunk 7 above)
Comment (tangjj11) [reply clang/lib/DPCT/DPCT.cpp:747]: done

File: clang/test/dpct/query_api_mapping/Math/test-after11.6.cu (hunk 10 above)
Comment (tangjj11) [reply clang/test/dpct/query_api_mapping/Math/test-after11.6.cu]: done

File: clang/lib/DPCT/DPCT.cpp (hunk 6 above)
Comment (tangjj11) [reply clang/lib/DPCT/DPCT.cpp]: done

File: clang/lib/DPCT/DPCT.cpp (hunk 9 above)
Comment (tangjj11) [reply clang/lib/DPCT/DPCT.cpp]: done

File: clang/test/dpct/query_api_mapping/Math/test-after11.6.cu (hunk 11 above)
Comment (tangjj11) [reply clang/test/dpct/query_api_mapping/Math/test-after11.6.cu:12]: Cannot add lit test, should add behavior test.

**Review by ziranzha — APPROVED** [review]

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**